vmdb/db/fixtures/ae_datastore/ManageIQ/Service/Provisioning/StateMachines/Methods.class/__methods__/catalogiteminitialization.rb

```diff
@@ -1,126 +1,243 @@
+# CatalogItemInitialization.rb
 #
-# Description: Tag service and set provison options based on service dialog entries.
-# 1. Look for all Service Dialog Options in the service_template_provision_task.dialog_options
-#    (i.e. Dialog options that came from either a Catalog Bundle Service or a Service Dialog)
-# 2. Service dialog option keys that match the regular expression /^tag_\d*_(.*)/i
-#    (I.e. <tag>_<sequence_id>_variable,  tag_0_function, tag_1_environment) will be used to
-#    tag the destination Catalog Item Service and any subordinate miq_provision tasks
-# 3. The remaining Service Dialog Option keys are simply passed into the subordinate
-#    miq_provision object. I.e. option_0_vm_memory => 2048
+# Author: Kevin Morey <kmorey@redhat.com>
+# License: GPL v3
 #
-# Inputs: $evm.root['service_template_provision_task'].dialog_options
+# Description: This method Performs the following functions:
+# 1. YAML load the Service Dialog Options from @task.get_option(:parsed_dialog_options))
+# 2. Set the name of the service
+# 3. Set tags on the service
+# 5. Override miq_provision task with any options and tags
 #
+begin
+  def log_and_update_message(level, msg, update_message = false)
+    $evm.log(level, "#{msg}")
+    @task.message = msg if @task && (update_message || level == 'error')
+  end
 
-# Look for service dialog variables in the dialog options hash that start with "tag_[0-9]",
-def get_tags_hash(dialog_options)
-  # Setup regular expression for service dialog tags
-  tags_regex = /^dialog_tag_\d*_(.*)/
+  # Loop through all tags from the dialog and create the categories and tags automatically
+  def create_tags(category, single_value, tag)
+    # Convert to lower case and replace all non-word characters with underscores
+    category_name = category.to_s.downcase.gsub(/\W/, '_')
+    tag_name = tag.to_s.downcase.gsub(/\W/, '_')
+    # if the category exists else create it
+    unless $evm.execute('category_exists?', category_name)
+      log_and_update_message(:info, "Creating Category: {#{category_name} => #{category}}")
+      $evm.execute('category_create', :name         => category_name,
+                                      :single_value => single_value,
+                                      :description  => "#{category}")
+    end
+    # if the tag exists else create it
+    return if $evm.execute('tag_exists?', category_name, tag_name)
+    log_and_update_message(:info, "Creating tag: {#{tag_name} => #{tag}}")
+    $evm.execute('tag_create', category_name, :name => tag_name, :description => "#{tag}")
+  end
 
-  tags_hash = {}
+  def create_category_and_tags_if_necessary(dialog_tags_hash)
+    dialog_tags_hash.each do |category, tag|
+      if tag.kind_of?(Array)
```

Comment:
@tinaafitz 
Can we make this into 1 code block that works for Array as well as single item, like we did in other places